site/public/js/pdf/PDFInitToolbar.js

```diff
@@ -146,33 +147,101 @@ window.onbeforeunload = function() {
+            fd.append('annotation_layer', JSON.stringify(annotation_layer));
+            fd.append('GENERAL_INFORMATION', JSON.stringify(GENERAL_INFORMATION));
+            fd.append('csrf_token', csrfToken);
+            fd.append('pdf', pdf);
+            let url = buildCourseUrl(['gradeable', GENERAL_INFORMATION['gradeable_id'], 'pdf', 'annotations']);
+            localStorage.setItem('rotate', rotateVal);
+            render(window.GENERAL_INFORMATION.gradeable_id, window.GENERAL_INFORMATION.user_id, window.GENERAL_INFORMATION.grader_id, window.GENERAL_INFORMATION.file_name, window.GENERAL_INFORMATION.file_path);
+            $.ajax({
+                type: 'POST',
+                url: url,
+                data: fd,
+                processData: false,
+                contentType: false,
+                success: function(data){
+                    let response = JSON.parse(data);
+                    if(response.status == "success"){
+                        $('#save_status').text("Saved");
+                        $('#save_status').css('color', 'black');
+                    }
+                    else {
+                        alert(data.message);
+                    }
+                },
+                error: function(){
+                    alert("Something went wrong, please contact a administrator.");
                 }
-            },
-            error: function(){
-                alert("Something went wrong, please contact a administrator.");
-            }
-        });
+            });
+            zoom("custom", zoom_level);
+            document.getElementById("file_content").scrollLeft = sLeft;
+            document.getElementById("file_content").scrollTop = sTop;
+            return;
+        }
+        else {
+            pageContainer = document.getElementById("pageContainer"+i);
+            document.getElementById("file_content").scrollTop = pageContainer.offsetTop;
+            document.getElementById("file_content").scrollLeft = pageContainer.offsetLeft;
+            $(document).ready(function() {
+                html2canvas(pageContainer).then(function(canvas) {
+                        let ctx = canvas.getContext('2d');
+                        let annotationLayer = pageContainer.getElementsByClassName("annotationLayer")[0];
+                        let scale = annotationLayer.scale;
+                        let children = annotationLayer.children;
+                        ctx.scale(zoom_level/100, zoom_level/100);
+                            for (let j = 0; j < children.length; j++) {
+                                let annotation = children[j];
+                                if(annotation.getAttribute("data-pdf-annotate-type") == "drawing"){
+                                    let path2 = new Path2D(annotation.getAttribute("d"));
+                                    ctx.strokeStyle = annotation.getAttribute("stroke");
```

Comment:
It looks like the annotations are getting read from the annotation layer and then getting drawn onto the canvas which is getting converted to an image, is there a way to directly take the annotations and use pdfJS to bake the annotations back into the pdf?

Is this something we can use http://raw.githack.com/MrRio/jsPDF/master/docs/module-context2d.html ?